File: Leetcode_with_Cpp/654constructMaximumBinaryTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* constructMaximumBinaryTree(vector<int>& nums) {
        return build(nums, 0, nums.size() - 1);
    }
    TreeNode* build(vector<int>& nums, int lo, int hi) {
        if (lo > hi)
            return NULL;
        if (lo == hi)
            return new TreeNode(nums[lo]);
        int rootIndex = lo;
        for (int i = lo + 1; i <= hi; i++) {
            if (nums[i] > nums[rootIndex])
                rootIndex=i;
        }
        TreeNode* root = new TreeNode(nums[rootIndex]);
        root->left = build(nums, lo, rootIndex - 1);
        root->right = build(nums, rootIndex + 1, hi);
        return root;
    }
};

File: py/654constructMaximumBinaryTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def constructMaximumBinaryTree(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        if not nums:
            return None
        max_value = max(nums)
        idx = nums.index(max_value)
        root = TreeNode(max_value)
        root.left = self.constructMaximumBinaryTree(nums[:idx])
        root.right = self.constructMaximumBinaryTree(nums[idx + 1 :])
        return root